compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -310,6 +310,34 @@ pub fn unexpected_hidden_region_diagnostic(
     err
 }
 
+/// Structurally compares two types, modulo any inference variables.
+///
+/// Returns `true` if two types are equal, or if one type is an inference variable compatible
+/// with the other type. A TyVar inference type is compatible with any type, and an IntVar or
+/// FloatVar inference type are compatible with themselves or their concrete types (Int and
+/// Float types, respectively). When comparing two ADTs, these rules apply recursively.
+pub fn same_type_modulo_infer(a: Ty<'tcx>, b: Ty<'ctx>) -> bool {
+    match (&a.kind(), &b.kind()) {
+        (&ty::Adt(did_a, substs_a), &ty::Adt(did_b, substs_b)) => {
+            if did_a != did_b {
+                return false;
+            }
+
+            substs_a.types().zip(substs_b.types()).all(|(a, b)| same_type_modulo_infer(a, b))
+        }
+        (&ty::Int(_), &ty::Infer(ty::InferTy::IntVar(_)))
+        | (&ty::Infer(ty::InferTy::IntVar(_)), &ty::Int(_) | &ty::Infer(ty::InferTy::IntVar(_)))
+        | (&ty::Float(_), &ty::Infer(ty::InferTy::FloatVar(_)))
+        | (
+            &ty::Infer(ty::InferTy::FloatVar(_)),
+            &ty::Float(_) | &ty::Infer(ty::InferTy::FloatVar(_)),
+        )
+        | (&ty::Infer(ty::InferTy::TyVar(_)), _)
+        | (_, &ty::Infer(ty::InferTy::TyVar(_))) => true,
```

Comment:
I'm pretty sure there's code like this *somewhere*.

Reply:
Yeah, this is from [`fn equals(Ty, Ty) -> bool`](https://github.com/rust-lang/rust/blob/323a48d7b44f9b34a6cb67a6ea2105a155763119/compiler/rustc_infer/src/infer/error_reporting/mod.rs#L1102). 

We could possibly unify them by adding an `enum Mode { None, NumInfer, Infer }` or something. The complication is that this one cares about the non-Ty substs in Adt, where TyS::same_ty (and the function I added) doesn't...

Thoughts?

Reply:
I see. `fn equals` *has* to make sure the whole type are the same because this is used to determine whether it can be replaced with `_` in the output. It's fine.